src/tribler/core/components/ipv8/eva/transfer/tests/test_base.py

```diff
@@ -33,46 +52,75 @@ async def test_finish_double_call(transfer: Transfer):
 
     # The first call of the finish method should process exception and result
     # and clean the container property
-    transfer.finish()
+    await transfer.finish()
     assert transfer.finished
     assert not transfer.container
 
     # The second call should do nothing. To check correctness we have to set any
     # value to transfer.container and ensure that it will not bt changed.
     transfer.container = 'any'
-    transfer.finish()
+    await transfer.finish()
```

Comment:
**nitpick:**

In the new version of EVAProtocol, `transfer.finish` is asynchronous. That means the execution can switch to a different task in the middle of the `finish` method. Currently, the test checks only the case when two invocations of `task.finish` are called in sequence, one after the other. But, for example, the following sequence is also theoretically possible:

1. Someone calls `task.finish`
2. Task finish executes partially and suspends on `await self._release()` without executing `_release` yet.
3. The `terminate_by_timeout` task (that was not canceled yet) resumes and decides that it is time to finish the same task by timeout, so it calls `asyncio.create_task(self.finish(exception=exception))`

Currently, the double execution of `finish` seems to be safe, as, inside the method, you check not only `self.finished` but also `self.future.done()`, which should return True during the second invocation of `finish`. But in general, it looks safer to have a flag `task.finishing` that is set to True not at the end of `task._release` but at the beginning of `task.finish` to prevent multiple parallel invocations.

Reply:
Fixed by making`transfer.finish()` synchronous.